### PR TITLE
[dtensor] implement shard dim change with alltoall

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124879
* __->__ #124872
* #124871

as titled, we implement a dedicated communication op to allow efficient
sharding dimension change using alltoall, to replace our previous
allgather + local chunk

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 1]
```diff
@@ -579,3 +579,50 @@ TORCH_LIBRARY(_c10d_functional_autograd, m) {
...
+  }
+  // allocate outputs
+  std::vector<at::Tensor> outputs;
+  for (int i = 0; i < group_size; i++) {
+    outputs.push_back(input.new_empty(output_sizes).contiguous());
+  }
+  auto work = group->alltoall(outputs, inputs);
+
+  work->wait();
+  // TODO: it's very tricky to get the current async behavior work for shard dim
```
Comment (yifuwang) [inline torch/csrc/distributed/c10d/Functional.cpp:614]: We can certainly add this. Curious how desirable it is in this case?

**Review by XilunWu — APPROVED** [review]
Thanks for providing a clean way to enable `Shard(i)` to `Shard(j)` resharding using all-to-all on 1-d mesh. A good starting point to generalize this to n-d mesh.

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/distributed/c10d/Functional.cpp:614]: Currently this op is always sync, but if we can enable in async that would potentially help the case of eager mode reordering. I couldn't figure out how to do this cleanly yet so would appreciate your help here!

Essentially if the `gather_dim` is not 0, then we would need to torch.cat on the `outputs` on the `gather_dim`, but we cannot simply associate the `cat` output with the async work handle, as the `torch.cat` result is a new buffer than the output list

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 2]
```diff
@@ -579,3 +579,50 @@ TORCH_LIBRARY(_c10d_functional_autograd, m) {
...
+                 << output_sizes[shard_dim]
+                 << ") is not divisible by the group size (" << group_size
+                 << ").";
+  }
+  output_sizes[shard_dim] = output_sizes[shard_dim] / group_size;
+  std::vector<at::Tensor> inputs;
+  auto length = output_sizes[shard_dim];
+  for (int i = 0; i < group_size; i++) {
+    inputs.push_back(input.narrow(shard_dim, i * length, length).contiguous());
+  }
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/csrc/distributed/c10d/Functional.cpp:605]: I think chunk under the hood calls this, the reason I use narrow directly is that I want to make sure every `input` is contiguous(), so we don't need to go through the chunked list again to convert them to contigous

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 1 above)
Comment (yifuwang) [reply torch/csrc/distributed/c10d/Functional.cpp:614]: Sounds good, I'll give it a try

**Review by yifuwang — APPROVED** [review]
LG!

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 3]
```diff
@@ -579,3 +579,50 @@ TORCH_LIBRARY(_c10d_functional_autograd, m) {
...
+    LOG(WARNING) << "The first dimension of the shard_dim_alltoall input ("
+                 << output_sizes[shard_dim]
+                 << ") is not divisible by the group size (" << group_size
+                 << ").";
+  }
+  output_sizes[shard_dim] = output_sizes[shard_dim] / group_size;
+  std::vector<at::Tensor> inputs;
+  auto length = output_sizes[shard_dim];
+  for (int i = 0; i < group_size; i++) {
+    inputs.push_back(input.narrow(shard_dim, i * length, length).contiguous());
```
Comment (bdhirsh) [inline torch/csrc/distributed/c10d/Functional.cpp:604]: just FYI (I got tagged on internal since I'm the current PT2 operator oncall) - this op is marked as pt2_compliant, but it doesn't look like it's implemented in a way that will work with dynamic shapes. Let me know if you want some pointers for fixing it!
